[DONE] [CR] Smoking rack interactions expansion

nexusmrsep opened the pull request:
### **What does it do?**

This PR changes how **smoking rack** works to resolve #24270 
Smoking rack new behavior is similar to charcoal kiln, where you load it with appropriate food and charcoal, light up and then leave smoking for enough time to do it's job.

### **Racionale:**
Smoking food by using recipes wastes precious time, forcing a watchful vigil over food being smoked.
This option allows to batch smoke food by placing it in the smoking rack, loading it with charcoal, firing it up, and leaving it behind while it does it's job.
Also see  #24270 for more info.

### **Features**

1. Smoking rack has an interactive menu with dynamic options (mainly to utilize existing reload_furniture function and to bypass NO_ITEM flag).
2. Smoking rack allows food with SMOKABLE flag, does't allow already SMOKED (flag) food, and other food 
3. It converts non-smoked food to smoked food (TODO: need a better way to pair smokable food with smoked products, other then hard-coding it, json-izing would be ideal option)
4. It holds max 20 liters of food, and needs 6 hours to finish smoking.
5. Added new recipes for raw versions of existing (smoked) sausages - they require less time, have good batch ratios, and are considered raw (spoil time, possible parasites), so they need smoking to get the desired properties of the final product.
6. [update] Smoking rack emits smoke (its a _smoking_ rack, could you expect anything else?; added `fake_smoke_plume` fake item to streamline this process)
7. [update] adding food to rack via a popup menu.
8. [update] added 'inspect rack' menu, mainly to inspect its contents

### **Implementation:**
Let's be real here, I'm not a pro so I scavenged as much good code as I could and stitched it together like a Frankenstein's monster. May not be pretty, but hopefully it'll be alive. It got more complicated then anticipated, but it's on a right path.

### **TODO**:
- [x] smoothing food inserting menu
- [x] testing & debugging 
- [x] Inspect contents menu option

Night-Pryanik commented on data/json/items/comestibles.json:
```diff
@@ -2624,9 +2646,30 @@
+    "quench": -1,
+    "healthy": -1,
+    "nutrition": 32,
+    "description": "A hefty raw long pork sausage that has been prepared for smoking.",
```
pork?

nexusmrsep replied:
[long pork](https://en.wiktionary.org/wiki/long_pork)

Rivet-the-Zombie replied:
We should probably add notes in there so that we don't have to keep explaining that old euphemism to people.

nexusmrsep replied:
I hope writing it down as 'long-pork' will help to recognize that it's something 'special'. I don't deny, I also thought it was either real pork or something along 'pulled pork' the first time I saw it. I'll add it to nearest commit.

Amariithynar replied:
Considering it uses as ingredient human flesh, I don't think it really needs to be specifically noted, no? That said, surprised that it's not common knowledge.

Night-Pryanik replied:
> That said, surprised that it's not common knowledge.

Not everyone here is a native English speaker.

Vasyan2006 commented:
Player should check and control this rack few times while cooking, otherwise the food can be burned or the fire will flame out. Real racks and charcoal kilns are observed during the entire process (unless you have AI multicoocker).
In addition this idle behaviour should be expanded to most cooking recepies. It is not really required to sit by the fire all the time if you want to boil water.

nexusmrsep commented:
> Player should check and control this rack few times while cooking

Such behavior would invalidate the whole reason of this PR as stated in #24270, as cooking actions demanding player's time are presently achieved by recipes. I'm not planning on doing that on this PR.

> It is not really required to sit by the fire all the time if you want to boil water.

True, but might be tedious to implement on a larger scale. There is however a delayed transformation action, that can be possibly applied to tea kettle, but I'm not sure that would be satisfactory. Other possible approach would be something along the code of fermenting vat. Just few loose ideas for someone who might want to code this.

sfsworms commented:
Will you also do something similar for the food dehydrator?

nexusmrsep commented:
> Will you also do something similar for the food dehydrator?

I'd love to, but I don't know if I got the skill. In my mind this PR will put some outlines for similar projects, but to not bloat the code with similar parts it would be ideal to have an universal mass item converter that would use json-ized matchmaking like:

 `{
 "type": "dehydrator",
 "pseudo_item": "dehydrator_active",
 "time": 12000,
 "products": [ [ "meat, "dehydrated meat" ],
                      [ "ingredient", "product" ],
                      [ "ingredient", "product" ] ]
}`

 coded in `iuse_actor.cpp` that would be called by specified item, so it would know what to convert and into what. Then and item-specific set of conditions could be coded separately that would be checked before conversion, etc. To do this by myself I'd have to learn more about C++, as  dynamic code used in `iuse_actor` is above my level at this moment.

Perhaps in the future if no one else will, I'd try.

nexusmrsep commented:
Added emission of smoke from the smoking rack, and for this purpose a fake smoke item (smoke generator) . It'll be fun when people used to having smoke racks indoors will activate them.

nexusmrsep commented:
Jenkins rebuild

Amariithynar commented:
@nexusmrsep could you have "smoked meat" be able to be added to the smoking rack to then convert to "dehydrated meat"?

nexusmrsep commented:
Technicaly - yes. That is actually a fair point, since `char_smoker` is used in many dehydrated food recipes. Consider it done.

Night-Pryanik commented:
> a fake smoke item

I think adding new fake items should be avoided at all times unless absolutely necessary. You can get this `smoke` item when smoking rack being in the middle of smoking process will be destroyed or deconstructed.

DracoGriffin commented:
>I think adding new fake items should be avoided at all times unless absolutely necessary. You can get this smoke item when smoking rack being in the middle of smoking process will be destroyed or deconstructed.

I think nexus is trying to [replicate the smoke](https://github.com/CleverRaven/Cataclysm-DDA/blob/64458e14447f1942ecf9ff8ccd9ff5a50eaf90b6/src/ballistics.cpp#L245) from [flintlock weapons](https://github.com/CleverRaven/Cataclysm-DDA/blob/f8346fd50b24d71df0715339ab0d091b7b24a8e8/data/json/items/ammo/flintlock.json#L21) but applied to smoking racks, whether it was done appropriately or not, I don't know.

Actually, probably not, looking at it a bit more closer.

Nexus, were you aware of a similar effect in the links above? Maybe refactor?

nexusmrsep commented:
> You can get this smoke item when smoking rack being in the middle of smoking process will be destroyed or deconstructed.

I'm well aware of that so `item::process_fake_smoke`  takes care of that, as it checks if the fake item is still in existing 'f_smoking_rack_active`. Also as `fake_smoke_plume' is on a countdown equal to smoking time so it fades away when process id done all by itself, and triggers 'on_smoke_out' in 'iexamine` to finalize smoking rack products.

nexusmrsep commented:
@Amariithynar  I've added the ability to batch dehydrate food in the smoking rack, with that addition of suggested mechanic: meat --> smoked meat --> dehydrated meat. Also for other applicable products from 'dry' list, but without 'smoked' phase if they had no such.

FYI I'm on a slow, but hopefully steady process of testing, debugging, and polishing this hefty piece of code it developed to. I believe I'm more towards the end then the beginning, but those menus can be tricky if you try to use existing functions and don't build everything from the scratch, and I met many problems with achieving compatibility.

nexusmrsep commented:
Update: It's working. I've cleaned up the menus, removed various bugs and erratic behavior. One more idea to implement: 'inspect contents' menu option to check what's currently inside. And another round of tests to confirm everything works fine. If nothing comes up ETA is 1-2 days.

nexusmrsep commented:
Ok after a few series of testing, I consider it done at this point. It there are some extra fixes to be done, it can be done on the go, or later.

nexusmrsep commented:
Also as for `fake_smoke_plume` I changed its name to "smoldering embers" with appropriate description, so if you disassemble the rack and manage to see them (for a brief moment before `process` kills them) you won't be confused of what you see. Same happens with other ways to destroy the rack, for ex. via bashing, or rampaging vehicle --> embers are destroyed at the nearest `process` update.

Mecares commented:
I would really like this change expanded to the pottery recipes and brick crafting.

nexusmrsep commented:
I'd certainly like to try to do that. But I'd rather json-ize this PR first, and make it generalized so it could do different things with same code, so this has to wait a bit.

cainiaowu commented on data/json/furniture.json:
```diff
@@ -2655,7 +2655,34 @@
+    "type" : "furniture",
+    "id" : "f_smoking_rack_active",
+    "name": "active smoking rack",
+    "description": "A special rack designed to smoke food for better preservation and taste.",
```
Maybe add one additional line like"it is currently lit" to inform the player better?

nexusmrsep replied:
Ok. Asap.

cainiaowu commented on data/json/items/fake.json:
```diff
@@ -102,5 +102,14 @@
+        "copy-from": "fake_item",
+        "type": "GENERIC",
+        "name": "smoldering embers",
+        "description": "A handful of smoldering embers emitting smoke.  They are fading away even when you look at them.",
```
I am not sure if the player can see this item when playing. Perhaps add some line like "this is a bug if you see it" to inform the player?

nexusmrsep replied:
If a player destroyes the active rack there is a sliver of time before process() kills the embers where it can be seen. In this regard I have not decided to do that, rather I want the player to see current description that suggests that they wil disappear instantly.

cainiaowu commented on data/json/items/comestibles/carnivore.json:
```diff
@@ -246,7 +249,7 @@
     "fun": 5,
     "proportional": { "price": 6, "calories": 0.95 },
-    "flags": [ "EATEN_HOT" ]
+    "flags": [ "EATEN_HOT", "SMOKED", "SMOKABLE" ]
```
Is this SMOKABLE flag added by accident? Accepted and not accepted at the same time.

nexusmrsep replied:
No, not by accident. It follows this stages: meat-->smoked meat-->dehydrated meat. The middle stage is both smokable and smoked. Code itself allows it to be smoked further.

nexusmrsep commented:
@cainiaowu  your review gave me a hint to remove the ability to deconstruct the rack while it's active (one less chance for the player to see the embers) and while doing that I found and removed a bug with inserting rotten food into the rack. Thanks.

nexusmrsep commented:
Jenkins rebuild

nexusmrsep commented:
@DracoGriffin  First of all - thanks for testing.

> Crashes when removing any amount of charcoal or food stored in smoking rack, even after successfully lighting up smoking rack.

Can't reproduce it in my compiled version. Maybe it was the broken Mannwurst - that i've fixed (it was result of not-so-well-done conflict resolving originating from copy-from PR #24554).
Please try again, and if it crashes this time too, tell me how to recreate it, step by step.

> (...) there should probably be a limit on charcoal (...)

Doable, but what would be the limit?

> (...) it continues to accept more foodstuff even after it's started smoking

It's a bug, and it was squished. Good call.

Also found a a bug with stacked rotten and fresh food resulting in inserting rotten food, despite using filters. It seems that on few occasions when player, or location, holds both fresh and rotten food of the same type at the same time, even with filters used, the later use of 'consume_items()' does't care anymore and uses whatever comes first. Had to use a non-optimal hack to bypass the problem, since I have no idea how to do it in a better way. It forces player to get rid of rotten food first manually, but prevents inserting & smoking rotten food.

I have also coded those features:
- removing charcoal no longer stops the smoking process, as you remove only excess charcoal;
- "Quench burning charcoal" option appears in menu when rack is active & smoking, so you can stop the process.

Night-Pryanik commented on src/iexamine.cpp:
```diff
@@ -3865,10 +3873,29 @@ void smoker_load_food( player &p, const tripoint &examp )
     comp_selection<item_comp> selected = p.select_item_component( comps, 1, inv, true );
     moved = p.consume_items( selected, 1 );
+
+    // hack, becouse consume_items doesn't seem to care of what item is consumed despite filters
```
Because.

Brambor commented on data/json/items/comestibles.json:
```diff
@@ -2528,9 +2550,18 @@
     "copy-from": "sausage",
     "name": "Mannwurst",
+    "flags": [ "EATEN_HOT", "SMOKED" ],
     "description": "A hefty long pork sausage that has been cured for long term storage.  Very tasty, if you're in the market for human flesh.",
```
_"cured"_ could be replaced with _"cured and smoked"_ to be consistent with sausage.
